lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,169 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module Dsc
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        @output_ready_event = self.class.create_event(@output_ready_event_name)
+
+        @stdin, @stdout, @ps_process = Open3.popen2(cmd)
+        @stdin.sync = true
+        @stdout.sync = true
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code)
+        # always need a trailing newline to ensure PowerShell parses code
+        out = exec_read_result(<<-CODE
+        $event = [Threading.EventWaitHandle]::OpenExisting("#{@output_ready_event_name}")
+
+        $Error.Clear()
+        $LASTEXITCODE = 0
+
+        #{powershell_code}
+
+        [Void]$event.Set()
```

Comment:
If `#{powershell_code}` fails to parse or execute for some reason, it's possible the event never gets set here.  If it never gets set, Puppet will deadlock.

We should carefully review the parameters added to the Invoke-DscResource PowerShell template to ensure that there's no opportunity for injection here as well.